internal/uvm/scsi.go

```diff
@@ -332,13 +357,160 @@ func (uvm *UtilityVM) GetScsiUvmPath(ctx context.Context, hostPath string) (stri
+// the uvm `vm`. If `sm` is read only then it is simply added to the `vm`. But if it is a
+// writeable mount(e.g a scratch layer) then a copy of it is made and that copy is added
+// to the `vm`.
+func (sm *SCSIMount) Clone(ctx context.Context, vm *UtilityVM, cd *cloneData) error {
+	var (
+		dstVhdPath string = sm.HostPath
+		err        error
+		dir        string
+		conStr     string = fmt.Sprintf("%d", sm.Controller)
+		lunStr     string = fmt.Sprintf("%d", sm.LUN)
+	)
+
+	if !sm.readOnly {
+		// Copy this scsi disk
+		// TODO(ambarve): This is a writeable SCSI mount. It can either be the
+		// scratch VHD of the UVM or it can be a SCSI mount that belongs to some
+		// container which is being automatically cloned here as a part of UVM
+		// cloning process. We will receive a request for creation of this
+		// container later on which will specify the storage path for this
+		// container.  However, that storage location is not available now so we
+		// just use the storage of the uvm instead. Find a better way for handling
+		// this. Problem with this approach is that the scratch VHD of the container
+		// will not be automatically cleaned after container exits. It will stay
+		// there as long as the UVM keeps running.
+
+		// For the scratch VHD of the VM (always attached at Controller:0, LUN:0)
+		// clone it in the scratch folder
+		dir = cd.scratchFolder
+		if sm.Controller != 0 || sm.LUN != 0 {
+			dir, err = ioutil.TempDir(cd.scratchFolder, fmt.Sprintf("clone-mount-%d-%d", sm.Controller, sm.LUN))
+			if err != nil {
+				return fmt.Errorf("error while creating directory for scsi mounts of clone vm: %s", err)
+			}
+		}
+
+		// copy the VHDX
+		dstVhdPath = filepath.Join(dir, filepath.Base(sm.HostPath))
+		log.G(ctx).WithFields(logrus.Fields{
+			"source hostPath":      sm.HostPath,
+			"controller":           sm.Controller,
+			"LUN":                  sm.LUN,
+			"destination hostPath": dstVhdPath,
+		}).Debug("Creating a clone of SCSI mount")
+
+		if err = copyfile.CopyFile(ctx, sm.HostPath, dstVhdPath, true); err != nil {
+			return err
+		}
+
+		if err = grantAccess(ctx, cd.uvmID, dstVhdPath, VMAccessTypeIndividual); err != nil {
+			os.Remove(dstVhdPath)
+			return err
+		}
+	}
+
+	if cd.doc.VirtualMachine.Devices.Scsi == nil {
+		cd.doc.VirtualMachine.Devices.Scsi = map[string]hcsschema.Scsi{}
+	}
+
+	if _, ok := cd.doc.VirtualMachine.Devices.Scsi[conStr]; !ok {
+		cd.doc.VirtualMachine.Devices.Scsi[conStr] = hcsschema.Scsi{
+			Attachments: map[string]hcsschema.Attachment{},
+		}
+	}
+
+	cd.doc.VirtualMachine.Devices.Scsi[conStr].Attachments[lunStr] = hcsschema.Attachment{
+		Path:  dstVhdPath,
+		Type_: sm.attachmentType,
+	}
+
+	clonedScsiMount := newSCSIMount(vm, dstVhdPath, sm.UVMPath, sm.attachmentType, 1, sm.Controller, sm.LUN, sm.readOnly)
+
+	vm.scsiLocations[sm.Controller][sm.LUN] = clonedScsiMount
+
+	return nil
+}
+
+func (sm *SCSIMount) GetSerialVersionID() uint32 {
+	return scsiCurrentSerialVersionID
```

Comment:
Should this return `sm.serialVersionID` instead?

Reply:
This doesn't really matter as of now because this function is not called anywhere. It is just there to make sure that the resource implements clone interface and hence handles serialization versions. We can update this later if we actually call this function somewhere. (Same for vsmb)